### ISSUE TITLE
Unify all CellManagers

### ISSUE THREAD

**Issue by ed255** [privacy-scaling-explorations/zkevm-circuits#1203]
Currently we have a CellManager in EVM Circuit that grows by rows, a CellManager in Keccack Circuit that grows by columns, and soon we will have a new one in MPT Circuit.

Try to unify them by having a single type that can be configured to work one way or the other, or maybe define a trait and implement two CellManagers.

@Brechtpd let us know if you're interested in working on this (or maybe you're already working on it?).

**Comment by Brechtpd** [comment]
I will be making sure that the MPT circuit and the EVM circuit will be able to use the same CellManager, the overall design of both these circuits will be quite similar.

The keccak circuit is significantly different where pretty much the only thing that is currently shared between both implementations is the name and the overall idea. Still some reuseability possible for sure, but currently not looking into doing this yet. I am interested in doing this but I cannot promise anything yet, just focusing on MPT + EVM for now.

**Comment by leolara** [comment]
I can work on this, but should we wait for the MPT to be merged? 

There are only two cell managers, keccak and evm currently, right?

**Comment by Brechtpd** [comment]
I don't think you need to wait on MPT as the goal is to keep changes to the EVM one to a minimum, with likely some extra functionality. I think it will be very useful to figure out how the Keccack and EVM one could be shared best regardless of future improvements.

I think just those two currently yes.

**Comment by aguzmant103** [comment]
Scope:
- Keep similar interface for the library user
- Unify both implementations
- Improve internal data structure and readability
- Optional: works for future circuits? (or just Keccak + XXXXXXX)

**Comment by leolara** [comment]
Proposal: https://hackmd.io/@leolara/zkEVM-Cell-Manager-refactoring/edit

**Comment by ed255** [comment]
@CPerezz in the past you had some ideas on how the CellManager unification should work, would you like to review Leo's proposal?

**Comment by CPerezz** [comment]
> @CPerezz in the past you had some ideas on how the CellManager unification should work, would you like to review Leo's proposal?

Sure! Will take a look during the weekend!

**Comment by ChihChengLiang** [comment]
I noticed a thread in @leolara's proposal got unanswered from almost everyone. Let me bring them back here, so that it's easier to get notified and answered.

The topic was whether to allow access of a particular row cell in FixedHeightStrategy. 

> @Brechtpd: It's sometimes necessary to place cells in some fixed layout over multiple rows for optimizations. For example of you have 2 rows available and you need to do 2 lookups in the same table, it's more efficient to place the data necessary to do those 2 lookups in the same location on those two rows, what way you only need to do a single lookup/row. This semi-manual management of the data is nice to have, I don't know how to best abstract that away (some kind of abstract region or column concept).
>> @leolara: For the lookups I think  Lookup(Table) CellType solves this. Am I right? 

> @adria0: I've been thinking in terms of table as memory and groups of cells like aligned words, meaning that you want to force some kind of basic layout to preserve row consistency.
> 
> Having something like: 
> 
> aligment_info=query_cells_aligned(&prev_aligment_info, &[cells_types]) returns a set of contiguous columns in the same row

> @hero78119: I am also thinking about an improvement related to this discussion threads. Let say a constraint q_sel x gate(cell_A, cell_B). And constraint is used in multiple rows, let say in another row it might be q_sel' x gate(cell_A', cell_B'), which turns out into 2 different constraints. 
> 
> If cell manager able to identify cell_A' and make it align with cell_A in same position, then above 2 constraint can be optimized into single constaint in final linearisation (q_sel + q_sel') * gate(cell_A' cell_B)

**Comment by leolara** [comment]
I would not enter into changes that require huge changes to the interface as this would require to change the code of the circuit itself. If we have ideas about things we could add to the CM interface, I would do that in separate tasks.

**Comment by aguzmant103** [comment]
Next steps: @ed255  and other people to take a look at this proposal in the upcoming week. Then make a decision to move forward

**Comment by ed255** [comment]
I have reviewed the design document.  Looks good to me! I like the overall idea of having a trait to describe a Strategy, and a wrapper over it that implements the cell_manager :fire: 

Here are some small comments to consider during the implementation, but nothing serious!  From my part the proposal looks good and the implementation can start any time.

For `FixedWidthStrategy` I think it's more intuitive to store:
```rust
struct FixedWidthStrategy {
    next: HashMap<CellType, (usize, usize)>
}
```
Where next is (column_index, rotation) of the next available cell.  On query you
do something like this:
```rust
let (next_column_index, next_rotation) = self.next;
self.next.0 += 1;
if self.next.0 == max_type_columns {
	self.next.0 = 0;
	self.next.1 += 1;
}
```
Maybe in `FixedHeightStrategy` you can do something similar.

Keccak uses `query_cell` without type, but there are not a lot of calls to this method, so I think it's ok to generalize to take a `cell_type`.

**Comment by ChihChengLiang** [comment]
@ed255 @leolara, What's the next step for this?

**Comment by ed255** [comment]
I think the next step is for @leolara to implement the changes following the proposal and send them in a PR :)

**Comment by leolara** [comment]
Yes, I started las week and sent a draft PR

**Comment by leolara** [comment]
Draft PR: https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1390

**Comment by leolara** [comment]
Yes, another PR for the keccak
